libraries/api-extractor-model/src/model/SourceLocation.ts

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as path from 'path';
+
+/**
+ * Constructor options for `SourceLocation`.
+ * @public
+ */
+export interface ISourceLocationOptions {
+  /**
+   * The project folder URL as defined by the `api-extractor.json` config `projectFolderUrl`
+   * setting.
+   */
+  projectFolderUrl?: string;
+
+  /**
+   * The file URL path relative to the `projectFolder` and `projectFolderURL` fields as
+   * defined in the `api-extractor.json` config.
+   */
+  fileUrlPath?: string;
+}
+
+/**
+ * The source location where a given API item is declared.
+ *
+ * @remarks
+ * The source location points to the `.ts` source file where the API item was originally
+   declared. However, in some cases, if source map resolution fails, it falls back to pointing
+   to the `.d.ts` file instead.
+ *
+ * @public
+ */
+export class SourceLocation {
+  private readonly _projectFolderUrl?: string;
+  private readonly _fileUrlPath?: string;
+
+  public constructor(options: ISourceLocationOptions) {
+    this._projectFolderUrl = options.projectFolderUrl;
+    this._fileUrlPath = options.fileUrlPath;
+  }
+
+  /**
+   * Returns the file URL to the given source location. Returns `undefined` if the file URL
+   * cannot be determined.
+   */
+  public get fileUrl(): string | undefined {
+    if (this._projectFolderUrl === undefined || this._fileUrlPath === undefined) {
+      return undefined;
+    }
+
+    return path.join(this._projectFolderUrl, this._fileUrlPath);
```

Comment:
Note that `path.join` doesn't quite work here, it shouldn't be used to construct URLs. For example, it will turn `"http://something"` into `"http:/something"`.

But simple concatenation doesn't work in all cases either, because `this._fileUrlPath` can include `"../"` segments (see the api-extractor-scenarios bundledPackages test scenario).

Reply:
Would the [`URL`](https://developer.mozilla.org/en-US/docs/Web/API/URL/URL) constructor work here?

It should solve the outlined issue, e.g. `new URL("../path", "http://something/abc").href` => `http://something.com/path`

Reply:
I'm not sure if it quite solves the issue above. `new URL("path", "http://something.com/abc").href` => `"http://something.com/path"`, but we'd want `"http://something.com/abc/path"`.

`this. _fileUrlPath` is always relative to the `<projectFolder>`. Thus, it should only include `"../"` segments if it's pointing to a file outside of `<projectFolder>`. I'm wondering if maybe this can only happen if it's pointing to a `.d.ts` file - as in the case of the bundledPackages test scenario. If that's the case, then perhaps `SourceLocation.fileUrl` can return `undefined` for `.d.ts` files, as I don't believe they're typically checked into source control, so a constructed URL wouldn't be valid anyway? Then we can just use simple concatenation here.

@octogonz - what do you think?

Reply:
(Keep in mind that `new URL("Äpfel", "http://something.com/abc/").href` prints `http://something.com/abc/%C3%84pfel` )